src/handlers/diff_header_diff.rs

```diff
@@ -25,6 +26,15 @@ impl<'a> StateMachine<'a> {
         self.handle_pending_line_with_diff_name()?;
         self.handled_diff_header_header_line_file_pair = None;
         self.diff_line = self.line.clone();
+
+        // Fill header information from the "diff" line, may be overwritten later.
+        let name = get_repeated_file_path_from_diff_line(&self.diff_line).unwrap_or_default();
+        self.minus_file = name.clone();
+        self.plus_file = name.clone();
+        self.minus_file_event = FileEvent::Change;
+        self.plus_file_event = FileEvent::Change;
+        self.current_file_pair = Some((self.minus_file.clone(), self.plus_file.clone()));
```

Comment:
Hi @imbrish, thanks for working on this! I'll admit to being rusty on the details of how this all works. Would you (while it's fresh in your head) be able to add some explanatory comments to this PR regarding the implementation? For example here -- my first reaction as maintainer is "Hm, OK so why are we setting all these to implement a feature regarding binary files and yet these changes are not behind any sort of `is_binary` flag".

Reply:
Sure. This part is the crux of the fix. Diff of modified binary files does not have the `---`/`+++` or `deleted file mode`/`new file mode` lines, and thus the `minus_file` and `plus_file` were never updated for such diffs, which caused the path from the previous diff to be shown again (as in the example). As for why setting all these... I am not sure, I have just seen the same in https://github.com/dandavison/delta/blob/main/src/handlers/diff_header.rs#L168-L172 and re-used that. I have just realized that the code there could possibly be removed, as now we do the same thing here upfront. I will try that later and explain the reasoning in the comment. I am just unsure, why it wasn't done like that in the first place. Do you suspect what could have been the reason for the current complexity?

Reply:
Thanks!

> I am just unsure, why it wasn't done like that in the first place. Do you suspect what could have been the reason for the current complexity?

Hm, I don't have it all paged in but it looks like this PR added those lines: https://github.com/dandavison/delta/pull/945

Reply:
I have now tried removing the lines below, but realized they are still needed.

https://github.com/dandavison/delta/blob/084a6d19b2054f8778c375862f8e4ed3d9145120/src/handlers/diff_header.rs#L163-L182

This is because for both added and removed files, the diff line will be `diff --git a/foo b/foo` so these must be recognized by looking at the `new file mode` or `deleted file mode` lines.